### DISCUSSION TITLE
Behavior for [ (left-bracket)

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#109] [Polls]
### Interpretation semantics for `[` (left-bracket)

At the moment, the standard undefines interpretation semantics for the word [`[`](https://forth-standard.org/standard/core/Bracket). Hence, a system [may provide](https://forth-standard.org/standard/rationale#paragraph.A.3.4.3.2) any behavior when this word is encountered by the Forth text interpreter in interpretation state. In most systems this behavior is to do nothing. I.e. the _implemented_ interpretation semantics for this word are to do nothing.

One known historical exception from this majority is the cmForth system that raises a not found error in this case, since this word is defined in COMPILER word list only ([cmforth.fth#L409](https://github.com/ForthHub/cmFORTH/blob/master/cmforth.fth#L409)).

Can anybody provide other such examples?

```poll
question: What interpretation semantics for "[" do you prefer?
options:
  - To do nothing
  - To throw an exception
  - No preferences
```

**Comment by mgcaret** [comment]
OF816 implements a flag for “compile-only” words, which its [definition](https://github.com/mgcaret/of816/blob/master/asm/forth-dictionary.s#L1132) for ``[`` has.  Using a word with this flag throws an exception.  
 
I could remove the flag and the result would be effectively doing nothing (storing a zero in the state variable, which is already zero in the interpretation state).  Since using ``[`` in the interpretation state is likely an error, I prefer the exception.  If I really wanted to execute it, ``' [ EXECUTE`` would do the trick.

**Reply by ruv** [reply]
> using `[` in the interpretation state is likely an error

It's reasonable!

> If I really wanted to execute it, `' [ EXECUTE` would do the trick.

Yes, sure.  But a standard program cannot do it, since it isn't allowed to obtain the execution token for a word for which the standard doesn't define interpretation semantics (or execution semantics).

And if they all would be defined for `[`, then the phrase `' [ EXECUTE` shall be equivalent to `[` in interpretation state (see [Tick](https://forth-standard.org/standard/core/Tick)). That is, if the latter throws an exception, then the former shall do the same.

At the moment, a standard-compliant way to apply `execute` to this word (to its compilation semantics) is as follows:
```
[: postpone [ ;] execute
```
or even a more portable variant:
```
[: state @ if  postpone [  then ;] execute
```

-----

By the way, [`]`](https://forth-standard.org/standard/right-bracket) is an ordinary word (not immediate), but you seem [define it](https://github.com/mgcaret/of816/blob/master/asm/forth-dictionary.s#L1140) as immediate.

**Reply by mgcaret** [reply]
> By the way, ] is an ordinary word (not immediate), but you seem define it as immediate.

I don't seem to, I do.  It even says why in the comments.  OF816 is implemented follwing [IEEE 1275-1994](https://www.openbios.org/data/docs/of1275.pdf) (AKA "Open Firmware) and ``]`` is listed in "7.3.9.2.2 Immediate words," regardless of what any other standard might say.

**Reply by ruv** [reply]
I have created an [issue](https://github.com/mgcaret/of816/issues/11) concerning immediacy and Open Firmware in your repository.

**Comment by Bushmills** [comment]
As it happens did I just upload "yoda", a rather Forthish Forth-like source to bash functions compiler, which implements the same mechanism as cmForth by placing `[` into a compilation-context vocabulary. There is no equivalent in stateless or interpret only vocabularies, with the result that `[` will be determined an unknown word if attempted to reference in other than compilation state context.
Its repo can be found here: https://github.com/Bushmills/yoda
